content/number-theory/factor.h

```diff
@@ -1,52 +1,35 @@
 /**
- * Author: Lukas Polacek
- * Date: 2010-01-28
+ * Author: chilli, SJTU
+ * Date: 2019-04-24
  * License: CC0
- * Source: Wikipedia
- * Description: Pollard's rho algorithm. It is a probabilistic factorisation
- * algorithm, whose expected time complexity is good. Before you start using it,
- * run {\tt init(bits)}, where bits is the length of the numbers you use.
- * Returns factors of the input without duplicates.
- * Time: Expected running time should be good enough for 50-bit numbers.
+ * Source: https://github.com/FTRobbin/Dreadnought-Standard-Code-Library
+ * Description: Pollard's rho algorithm: a probabilistic factorisation
+ * algorithm whose expected time complexity is good.
+ * Time: On worst case numbers, $O(n^{1/4})$. Faster than that for most numbers.
  */
 #pragma once
 
 #include "ModMulLL.h"
 #include "MillerRabin.h"
-#include "eratosthenes.h"
 
-vector<ull> pr;
-ull f(ull a, ull n, ull &has) {
-	return (mod_mul(a, a, n) + has) % n;
-}
-vector<ull> factor(ull d) {
-	vector<ull> res;
-	for (int i = 0; i < sz(pr) && pr[i]*pr[i] <= d; i++)
-		if (d % pr[i] == 0) {
-			while (d % pr[i] == 0) d /= pr[i];
-			res.push_back(pr[i]);
-		}
-	//d is now a product of at most 2 primes.
-	if (d > 1) {
-		if (isPrime(d))
-			res.push_back(d);
-		else while (true) {
-			ull has = rand() % 2321 + 47;
-			ull x = 2, y = 2, c = 1;
-			for (; c==1; c = __gcd((y > x ? y - x : x - y), d)) {
-				x = f(x, d, has);
-				y = f(f(y, d, has), d, has);
-			}
-			if (c != d) {
-				res.push_back(c); d /= c;
-				if (d != c) res.push_back(d);
-				break;
-			}
-		}
+ull f(ull x, ull n) { return (mod_mul(x, x, n) + 1) % n; }
+ull Pollard(ull n) {
+	auto f = [&n](ull x) { return (mod_mul(x, x, n) + 1) % n; };
+	if (isPrime(n)) return n;
+	if (!(n & 1)) return 2;
+	for (ull i = 2;; i++) {
+		ull x = i, y = f(x), p = __gcd(n + y - x, n);
+		while (p == 1)
+			x = f(x), y = f(f(y)), p = __gcd(n + y - x, n);
+		if (p != n) return p;
 	}
-	return res;
+	return 0;
```

Comment:
no need for this given the inf loop

Reply:
Ah interesting. Didn't know that gcc wouldn't throw an error if it could tell it was an infinite loop.